distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Add_Row_Number.enso

```diff
@@ -0,0 +1,103 @@
+from Standard.Base import all
+
+import project.Data.Column.Column
+import project.Data.Column_Selector.Column_Selector
+import project.Data.Sort_Column.Sort_Column
+import project.Data.Set_Mode.Set_Mode
+import project.Data.Table.Table
+import project.Internal.Multi_Value_Key.Unordered_Multi_Value_Key
+import project.Internal.Multi_Value_Key.Ordered_Multi_Value_Key
+import project.Internal.Problem_Builder.Problem_Builder
+import project.Internal.Table_Helpers
+import project.Internal.Unique_Name_Strategy.Unique_Name_Strategy
+
+from project.Errors import Duplicate_Output_Column_Names
+from project.Internal.Java_Exports import make_long_builder
+
+polyglot java import org.enso.base.arrays.LongArray
+polyglot java import org.enso.table.data.column.storage.LongStorage
+
+## PRIVATE
+add_row_number : Table -> Text -> Integer -> Integer -> (Column_Selector | Vector Text) -> Vector (Text | Sort_Column) | Text | Sort_Column -> Problem_Behavior -> Table
+add_row_number table name from step group_by order_by on_problems =
+    problem_builder = Problem_Builder.new error_on_missing_columns=True
+    grouping_columns = table.columns_helper.select_columns_helper group_by True problem_builder
+    Unordered_Multi_Value_Key.validate_grouping_columns grouping_columns problem_builder
+    ordering = Table_Helpers.resolve_order_by table.columns order_by problem_builder
+    problem_builder.attach_problems_before on_problems <|
+        new_column = case ordering.is_empty of
+            True ->
+                case grouping_columns.is_empty of
+                    True -> make_range_column name from step table.row_count
+                    False -> make_grouped_enumeration name grouping_columns from step
+            False -> make_grouped_ordered_enumeration name grouping_columns ordering from step
+
+        column_names = table.column_names
+        renamed_table = if column_names.contains name . not then table else
+            problems = [Duplicate_Output_Column_Names.Error [name]]
+            on_problems.attach_problems_before problems <|
+                unique_name_strategy = Unique_Name_Strategy.new
+                unique_name_strategy.mark_used column_names
+                new_name = unique_name_strategy.make_unique name
+                new_columns = table.columns.map column->
+                    if column.name == name then column.rename new_name else column
+                Table.new new_columns
+        renamed_table.set new_column name set_mode=Set_Mode.Add
+
+## PRIVATE
+nth_index start step n =
+    start + n*step
+
+## PRIVATE
+make_range_column name start step length =
+    builder = make_long_builder length
+    0.up_to length . each ix->
```

Comment:
Why not use a Range here?

Reply:
We pass the created storage to the underlying Java methods. The Java Column needs a Java storage. Currently we just have `LongStorage` for integers.

I was thinking of creating a special storage for storing these ranges in O(1) space, but I thought it may be an overkill given that in other cases (when grouped or ordered) we need to use `LongStorage` anyway.

But since the default case is probably likely most common, maybe indeed it could be worth to implement the O(1) storage just for it.

Reply:
Perhaps, but I wasn't referring to that -- I just meant for the iteration. Instead of explicitly multiplying by step using nth_index, just create a Range with a step and iterate over that. It seems to be doing the same thing as Range with a step value.

Reply:
Fair, but I wanted to use the same logic to compute the nth enumeration element in all 3 'modes'.